csrc/scheduler/mma_utils.cpp

```diff
@@ -1859,11 +1876,20 @@ std::vector<ValGroup> canonicalDimOrdering(
   std::vector<ValGroup> ordering;
   ordering.reserve(
       batch_dims.size() + m_dims.size() + n_dims.size() + k_dims.size());
-  const auto insert = [&ordering](const VectorOfUniqueEntries<ValGroup>& v) {
+  const auto insert = [&ordering, &device_dims](
+                          const VectorOfUniqueEntries<ValGroup>& v,
+                          bool skip_device_dims = true) {
     for (auto it = v.vector().rbegin(); it != v.vector().rend(); ++it) {
-      ordering.push_back(*it);
+      const ValGroup& g = *it;
+      if (skip_device_dims && device_dims.has(g)) {
+        continue;
+      }
+      ordering.push_back(g);
     }
   };
+  // Insert the device dims first, then skip them when inserting dims from each
+  // other role
```

Comment:
Ensuring device dim groups are placed outside, even if that means we have non-consecutive dims within each role.